commons/protocol/src/main/scala/sbt/protocol/Protocol.scala

```diff
@@ -71,9 +73,71 @@ sealed trait ExecutionEngineEvent extends Event
 //              Requests (Reactive API)
 // ------------------------------------------
 
-final case class ClientInfo(uuid: String, configName: String, humanReadableName: String)
+// "protocolVersion" doesn't have semantics like major.minor;
+// versions are supposed to be backward compatible always (if they
+// weren't we'd need to do something more radical than send this,
+// pre-connection, such as change the sbt server launch
+// procedure). It's probably easier to use FeatureTag for most
+// "negotiation" scenarios, too. But you could also have a rule
+// like "in protocol version 3, if both sides are >= 3 you do
+// xyz."
+sealed abstract class ProtocolVersion protected (val name: String) {
+  final override def toString = name
+  final override def equals(x: Any): Boolean = x match {
+    case null => false
+    case other: ProtocolVersion => name == other.name
+    case _ => false
+  }
+  final override def hashCode(): Int = name.hashCode
+}
+object ProtocolVersion {
+  private implicit val resultToString = CanToString[ProtocolVersion](_.toString,
+    {
+      case "1" => ProtocolVersion1
+      // we don't want to explode here if we get a version
+      // we don't know. If it's handled somehow, it'd be
+      // elsewhere (in the client or server logic).
+      case other => ProtocolVersionUnknown
+    })
+
+  implicit val picklerUnpickler: Pickler[ProtocolVersion] with Unpickler[ProtocolVersion] =
+    canToStringPickler[ProtocolVersion]
+}
+
+case object ProtocolVersionUnknown extends ProtocolVersion("unknown")
+case object ProtocolVersion1 extends ProtocolVersion("1")
+
+// These enable or disable certain behaviors. For example you might have
+// a tag like "SupportsFoo" and if a client sets that tag the server
+// could enable "Foo", or whatever.
+sealed abstract class FeatureTag protected (val name: String) {
```

Comment:
FeatureTag and ProtocolVersion could obviously share a base class that would implement the common logic, but I couldn't make myself happy with the fact that the base class would be an implementation detail leaking into public API.